homeassistant/components/recorder/__init__.py

```diff
@@ -305,8 +305,82 @@ def _setup_connection(self):
         models.Base.metadata.create_all(self.engine)
         session_factory = sessionmaker(bind=self.engine)
         Session = scoped_session(session_factory)
+        self._check_schema()
         self.db_ready.set()
 
+    def _check_schema(self):
+        """Check if the schema needs to be upgraded."""
+        import homeassistant.components.recorder.models as models
+        schema_changes = models.SchemaChanges
+        current_version = getattr(Session.query(schema_changes).order_by(
+            schema_changes.change_id.desc()).first(), 'schema_version', None)
+
+        if current_version == models.SCHEMA_VERSION:
+            return
+        _LOGGER.debug("Schema version incorrect: %d", current_version)
+
+        if current_version is None:
+            current_version = self._inspect_schema_version()
+            _LOGGER.debug("No schema version found. Inspected version: %d",
+                          current_version)
+
+        for version in range(current_version, models.SCHEMA_VERSION):
+            new_version = version + 1
+            _LOGGER.info(
+                "Upgrading recorder db schema to version %d", new_version)
+            self._migrate_schema(new_version)
+            self._commit(schema_changes(schema_version=new_version))
+            _LOGGER.info(
+                "Upgraded recorder db schema to version %d", new_version)
+
+    def _migrate_schema(self, new_version):
+        """Perform operations to bring schema up to date."""
+        from sqlalchemy import Index, Table
+        import homeassistant.components.recorder.models as models
+
+        if new_version == 1:
+            def create_index(table_name, column_name):
+                """Create an index for the specified table and column."""
+                table = Table(table_name, models.Base.metadata)
+                index_name = "_".join(("ix", table_name, column_name))
+                index = Index(index_name, getattr(table.c, column_name))
+                _LOGGER.debug("Creating index for table %s column %s",
+                              table_name, column_name)
+                index.create(self.engine)
+                _LOGGER.debug("Index creation done for table %s column %s",
+                              table_name, column_name)
+
+            create_index("events", "time_fired")
+        else:
+            raise ValueError("No schema migration defined for version {}"
+                             .format(new_version))
+
+    def _inspect_schema_version(self):
```

Comment:
Exactly. This initial case is a bit of a conundrum. It would have been nice to have this from the start. The detection definitely doesn't ever need to be extended, and we can remove it eventually after a sufficient waiting period and assume all empty schema tables are new dbs.

I'm definitely open if you've got another idea, since this is pretty ugly, but this was the best I came up with to support migrations of existing dbs. This at least is only checked once for each installation.